src/d3a/d3a_core/sim_results/area_statistics.py

```diff
@@ -176,34 +190,45 @@ def _accumulate_house_trades(house, grid, accumulated_trades, past_market_types)
-
-    for market in getattr(grid, past_market_types):
-        for trade in market.trades:
-            if trade.buyer == house_IAA_name and trade.buyer != trade.offer.seller:
-                seller_id = area_name_from_area_or_iaa_name(trade.seller)
-                accumulated_trades[house.name]["consumedFrom"][seller_id] += trade.offer.energy
-                accumulated_trades[house.name]["spentTo"][seller_id] += trade.offer.price
-
-    for market in getattr(house, past_market_types):
-        for trade in market.trades:
-            if area_name_from_area_or_iaa_name(trade.seller) == house.name:
-                accumulated_trades[house.name]["consumedFromExternal"] += trade.offer.energy
-                accumulated_trades[house.name]["spentToExternal"] += trade.offer.price
-            elif area_name_from_area_or_iaa_name(trade.buyer) == house.name:
-                accumulated_trades[house.name]["producedForExternal"] -= trade.offer.energy
-                accumulated_trades[house.name]["earnedFromExternal"] += trade.offer.price
+    house_markets = getattr(house, past_market_types)
+    grid_markets = getattr(grid, past_market_types)
+    if house_markets is not None:
+        if type(house_markets) != list:
+            house_markets = [house_markets]
+        for market in house_markets:
+            for trade in market.trades:
+                if area_name_from_area_or_iaa_name(trade.seller) in child_names and \
+                        area_name_from_area_or_iaa_name(trade.buyer) in child_names:
+                    # House self-consumption trade
+                    accumulated_trades[house.name]["produced"] -= trade.offer.energy
+                    accumulated_trades[house.name]["earned"] += trade.offer.price
+                    accumulated_trades[house.name]["consumedFrom"][house.name] +=\
+                        trade.offer.energy
+                    accumulated_trades[house.name]["spentTo"][house.name] += trade.offer.price
+                elif trade.buyer == house_IAA_name:
+                    accumulated_trades[house.name]["earned"] += trade.offer.price
+                    accumulated_trades[house.name]["produced"] -= trade.offer.energy
+        for market in house_markets:
+            for trade in market.trades:
+                if area_name_from_area_or_iaa_name(trade.seller) == \
+                        house.name and trade.buyer not in child_names:
+                    accumulated_trades[house.name]["consumedFromExternal"] += trade.offer.energy
+                    accumulated_trades[house.name]["spentToExternal"] += trade.offer.price
+                elif area_name_from_area_or_iaa_name(trade.buyer) == \
+                        house.name and trade.seller not in child_names:
+                    accumulated_trades[house.name]["producedForExternal"] -= trade.offer.energy
+                    accumulated_trades[house.name]["earnedFromExternal"] += trade.offer.price
+
+    if grid_markets is not None:
```

Comment:
You can specify a default value when running `getattr()` (in line 194) and could save both if clauses here.

Reply:
I tired, but didn't.May be I can show you offline?